charms/worker/k8s/src/token_distributor.py

```diff
@@ -282,45 +346,49 @@ def allocate_tokens(
             token_strategy (TokenStrategy): The strategy of token creation.
             token_type (ClusterTokenType): The type of cluster token.
                 Defaults to ClusterTokenType.NONE.
-
-        Raises:
-            ValueError: If an invalid token_strategy is provided.
         """
-        revoke_on_join = token_strategy == TokenStrategy.CLUSTER
         units = relation.units
         if self.charm.app == relation.app:
             # include self in peer relations
             units |= {self.charm.unit}
         assert relation.app, f"Remote application doesn't exist on {relation.name}"  # nosec
 
         # Select the appropriate token creation strategy
-        token_strat = self.token_creation_strategies.get(token_strategy)
-        if not token_strat:
-            raise ValueError(f"Invalid token_strategy: {token_strategy}")
-
-        log.info("Allocating %s tokens", token_type.value)
-        status.add(ops.MaintenanceStatus(f"Allocating {token_type.value} tokens"))
+        tokenizer = self.token_strategies.get(token_strategy)
+        assert tokenizer, f"Invalid token_strategy: {token_strategy}"  # nosec
+
+        log.info("Allocating %s %s tokens", token_type.name.title(), token_strategy.name.title())
+        status.add(
+            ops.MaintenanceStatus(
+                f"Allocating {token_type.name.title()} {token_strategy.name.title()} tokens"
+            )
+        )
         local_cluster = self.charm.get_cluster_name()
         relation.data[self.charm.unit]["node-name"] = self.node_name
-        relation.data[self.charm.unit]["joined"] = local_cluster
         relation.data[self.charm.unit]["cluster-name"] = local_cluster
+        if not tokenizer.allocator_needs_tokens:
+            # the allocator doesn't need a token to join, mark as already joined
+            relation.data[self.charm.unit]["joined"] = local_cluster
+
```

Comment:
some tokens may need to be allocated on the "allocating node".  Depending on the strategy (cos or cluster)